### PR TITLE
[CSSimplify] Delay `inout` type to pointer conversion until `inout` is sufficiently resolved

### PR DESCRIPTION
**Pull request by xedin**
If left-hand side of any conversion constraint is `inout` type
and right is a pointer (or optional thereof), delay simplification
until `inout` is at least partially structurally resolved (cannot
be a type variable or dependent member) because eager simplification
won't record all of the possible conversions.

Allow inferring type of `inout` from a pointer type
(or optional thereof) but delay the binding set because
it might not be complete and object type of `inout` could
also be an Array or C-style pointer type.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

File: lib/Sema/CSBindings.cpp [hunk 1]
```diff
@@ -1484,6 +1484,23 @@ PotentialBindings::inferFromRelational(Constraint *constraint) {
         AdjacentVars.insert({typeVar, constraint});
     }
 
+    // Infer a binding from `inout $T <convertible to> Unsafe*Pointer<...>?`.
+    if (first->is<InOutType>() &&
+        first->getInOutObjectType()->isEqual(TypeVar)) {
+      if (auto pointerTy = second->lookThroughAllOptionalTypes()
+                               ->getAnyPointerElementType()) {
```
Comment (hamishknight) [inline lib/Sema/CSBindings.cpp:1491]: Nit: This should probably be named `pointeeType`

File: lib/Sema/CSBindings.cpp [hunk 2]
```diff
@@ -1484,6 +1484,23 @@ PotentialBindings::inferFromRelational(Constraint *constraint) {
...
+        if (!pointerTy->isTypeVariableOrMember()) {
+          // The binding is as a fallback in this case because $T could
+          // also be Array<X> or C-style pointer.
+          if (constraint->getKind() >= ConstraintKind::ArgumentConversion)
+            DelayedBy.push_back(constraint);
+
+          return PotentialBinding(pointerTy, AllowedBindingKind::Subtypes,
```
Comment (hamishknight) [inline lib/Sema/CSBindings.cpp:1498]: Could this use `AllowedBindingKind::Exact`?

**Review by hamishknight — APPROVED** [review]

File: lib/Sema/CSBindings.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSBindings.cpp:1498]: We can use that indeed!

File: lib/Sema/CSBindings.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSBindings.cpp]: ✅

File: lib/Sema/CSBindings.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSBindings.cpp]: ✅

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**